runtime/interpreter/deployedcontract.go

```diff
@@ -37,18 +41,62 @@ func NewDeployedContractValue(
 	name *StringValue,
 	code *ArrayValue,
 ) *SimpleCompositeValue {
+	publicTypesFuncValue := newPublicTypesFunctionValue(inter, address, name)
 	return NewSimpleCompositeValue(
 		inter,
 		sema.DeployedContractType.TypeID,
 		deployedContractStaticType,
 		deployedContractFieldNames,
 		map[string]Value{
-			sema.DeployedContractTypeAddressFieldName: address,
-			sema.DeployedContractTypeNameFieldName:    name,
-			sema.DeployedContractTypeCodeFieldName:    code,
+			sema.DeployedContractTypeAddressFieldName:        address,
+			sema.DeployedContractTypeNameFieldName:           name,
+			sema.DeployedContractTypeCodeFieldName:           code,
+			sema.DeployedContractTypePublicTypesFunctionName: publicTypesFuncValue,
 		},
 		nil,
 		nil,
 		nil,
 	)
 }
+
+func newPublicTypesFunctionValue(inter *Interpreter, addressValue AddressValue, name *StringValue) FunctionValue {
+	// public types only need to be computed once per contract
+	var once sync.Once
+	var publicTypes *ArrayValue
+
+	address := addressValue.ToAddress()
+	return NewHostFunctionValue(inter, func(inv Invocation) Value {
+		once.Do(func() {
+			innerInter := inv.Interpreter
+			contractLocation := common.NewAddressLocation(innerInter, address, name.Str)
+			// we're only looking at the contract as a whole, so no need to construct a nested path
+			qualifiedIdent := name.Str
+			typeID := common.NewTypeIDFromQualifiedName(innerInter, contractLocation, qualifiedIdent)
+			compositeType, err := innerInter.GetCompositeType(contractLocation, qualifiedIdent, typeID)
+			if err != nil {
+				panic(err)
+			}
+
+			nestedTypes := compositeType.NestedTypes
+			pair := nestedTypes.Oldest()
+			yieldNext := func() Value {
+				if pair == nil {
+					return nil
+				}
+				typeValue := TypeValue{Type: ConvertSemaToStaticType(innerInter, pair.Value)}
+				pair = pair.Next()
+				return typeValue
+			}
+
+			publicTypes = NewArrayValueWithIterator(
+				innerInter,
+				NewVariableSizedStaticType(innerInter, PrimitiveStaticTypeMetaType),
+				common.Address{0x0},
```

Comment:
Can just use the default value of `Address` (an `[8]byte`)

```suggestion
				common.Address{},
```